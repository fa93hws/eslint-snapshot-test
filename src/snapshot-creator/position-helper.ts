import { Linter } from '@typescript-eslint/experimental-utils/dist/ts-eslint';

export type Position = {
  line: {
    start: number;
    end: number | undefined;
  };
  column: {
    start: number;
    end: number | undefined;
  };
}

export class PositionHelper {
  public constructor(private readonly lines: readonly string[]) {
  }

  public static getPosition(result: Linter.LintMessage): Position {
    return {
      line: {
        // line starts from 1 in eslint
        start: result.line - 1,
        end: result.endLine ? result.endLine - 1 : result.endLine,
      },
      column: {
        // column starts from 1 in eslint
        start: result.column - 1,
        end: result.endColumn ? result.endColumn - 1 : result.endColumn,
      },
    };
  }

  public getColumnOnLine(rawPosition: Position) {
    return {
      start: rawPosition.column.start,
      end: rawPosition.column.end ?? this.lines[rawPosition.line.start].length,
    };
  }
}
