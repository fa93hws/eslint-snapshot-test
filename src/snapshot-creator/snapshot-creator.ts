import { RuleTesterConfig, Linter, RuleModule } from '@typescript-eslint/experimental-utils/dist/ts-eslint'
import { ErrorWorker } from './workers/error-marker';

export class SnapshotCreator {
  private readonly linter: Linter = new Linter();
  public constructor(private readonly config: RuleTesterConfig) {
    this.linter.defineParser(config.parser, require(config.parser));
  }

  public mark<TOption extends readonly any[]>({ code, ruleName, rule }: {
    code: string;
    ruleName: string;
    rule: RuleModule<any, TOption, any>
  }) {
    if (!this.linter.getRules().has(ruleName)) {
      this.linter.defineRule(ruleName, rule);
    }

    return new ErrorWorker<TOption>({
      code,
      ruleName,
      linter: this.linter,
      config: this.config,
    });
  }
}
