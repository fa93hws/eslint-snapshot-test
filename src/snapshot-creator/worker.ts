import { RuleModule, ValidTestCase } from '@typescript-eslint/experimental-utils/dist/ts-eslint'

export const enum WorkerType {
  MARK,
  FIX,
}

type TestConfig<TOption extends readonly any[]> = Omit<ValidTestCase<TOption>, 'code'>;

export class Worker<TOption extends readonly any[]> {
  private ruleConfig: TestConfig<TOption> = {};
  private rule?: RuleModule<any, TOption, any>;
  private readonly type: WorkerType;

  public constructor(type: WorkerType) {
    this.type = type;
  }

  public withConfig(config: TestConfig<TOption>) {
    this.ruleConfig = config;
    return this;
  }

  public onRule(rule: RuleModule<any, TOption, any>) {
    this.rule = rule;
    return this;
  }
}
