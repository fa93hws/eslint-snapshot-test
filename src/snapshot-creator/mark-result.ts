import { Linter } from '@typescript-eslint/experimental-utils/dist/ts-eslint';
import { PositionHelper } from './position-helper';

export type MarkedLine = {
  // To be inserted in the original code, start from 0
  afterLine: number;
  text: string;
};

export type MarkResultFn = (param: {
  lintResult: readonly Linter.LintMessage[];
  positionHelper: PositionHelper;
}) => MarkedLine[];

function drawWaveString({
  start: columnStart,
  end: columnEnd,
}: {
  start: number;
  end: number;
}) {
  const leadingSpaces = ' '.repeat(columnStart);
  const waveLength = Math.max(columnEnd - columnStart, 1);
  const waves = '~'.repeat(waveLength);

  return leadingSpaces + waves;
}

const sortFn = (a: MarkedLine, b: MarkedLine) => a.afterLine - b.afterLine;

export const markResult: MarkResultFn = ({ lintResult, positionHelper }) => {
  const result: MarkedLine[] = [];
  lintResult.forEach(r => {
    const range = PositionHelper.getRange(r);
    const positions = positionHelper.parsePosition(range);
    positions.forEach(position => {
      const waveString = drawWaveString(position.column);
      result.push({
        afterLine: position.line,
        text: `${waveString}    [${r.message}]`,
      });
    });
  });
  result.sort(sortFn);
  return result;
};
