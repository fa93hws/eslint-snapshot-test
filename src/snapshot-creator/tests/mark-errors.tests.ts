import { SnapshotCreator } from "../snapshot-creator";
const semi = require('eslint/lib/rules/semi');

describe('when marking error to snapshot', () => {
  const snapshotCreator = new SnapshotCreator({
    parserOptions: {
      ecmaVersion: 2017,
      sourceType: 'module',
    },
    parser: '@typescript-eslint/parser',
  });
  const code = 'const a = 1';

  it('should generate the snapshot correctly', () => {
    snapshotCreator.mark(code)
    .onRule(semi)
    .withOptions(['error'])
    .render();

    expect(1).toBe(1);
  });
});
