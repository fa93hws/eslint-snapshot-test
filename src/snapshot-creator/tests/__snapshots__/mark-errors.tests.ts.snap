// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when marking error to snapshot generates snapshot for multiple one line error 1`] = `
"
var a = 1;
    ~    ['a' is assigned a value but never used.]
var b = 1;
var foo = 1;
    ~~~    ['foo' is assigned a value but never used.]
fn(b);"
`;

exports[`when marking error to snapshot generates snapshot for one line code 1`] = `
"
var a = 1;
    ~    ['a' is assigned a value but never used.]"
`;

exports[`when marking error to snapshot generates snapshot for the error at last column 1`] = `
"
var a = 1
         ~    [Missing semicolon.]"
`;

exports[`when marking error to snapshot generates snapshot with rule option 1`] = `
"
a = 1;
var b = 2;
    ~    ['b' is assigned a value but never used.]"
`;
